java/oauth-custom-challengehandler/src/main/java/com/esri/arcgisruntime/sample/oauthcustomchallengehandler/MainActivity.java

```diff
@@ -0,0 +1,308 @@
+
+  private static OAuthChallengeCountDownLatch sOAuthChallengeCountDownLatch;
+
+  private OAuthConfiguration mOAuthConfiguration = null;
+
+  private Dialog mOAuthLoginDialog;
+
+  @Override
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_main);
+
+    // set custom challenge handler on the AuthenticationManager
+    AuthenticationManager.setAuthenticationChallengeHandler(customAuthenticationChallengeHandler);
+
+    // initialize the dialog to display the OAuth login page
+    mOAuthLoginDialog = new Dialog(this);
+
+    // inflate MapView from layout
+    mMapView = (MapView) findViewById(R.id.mapView);
+
+    // add the OAuth configuration to the AuthenticationManager
+    try {
+      OAuthConfiguration config = new OAuthConfiguration(getString(R.string.portal_url), getString(R.string.client_id), getString(R.string.redirect_url));
+      AuthenticationManager.addOAuthConfiguration(config);
+    } catch (MalformedURLException e) {
+      throw new IllegalArgumentException(e.getMessage(), e.getCause());
+    }
+
+    // create a portal instance for the portal to load
+    Portal portal = new Portal(getString(R.string.portal_url));
+
+    // create a portal item with the itemId of the web map
+    PortalItem webMapItem = new PortalItem(portal, getString(R.string.webmap_id));
+
+    // create a map with the portal item
+    ArcGISMap map = new ArcGISMap(webMapItem);
+
+    // set the map to the map view
+    mMapView.setMap(map);
+  }
+
+  /**
+   * Anonymous class to define our customAuthenticationChallengeHandler.
+   */
+  AuthenticationChallengeHandler customAuthenticationChallengeHandler = new AuthenticationChallengeHandler() {
+    /**
+     * Handles the incoming AuthenticationChallenge, returning a response that contains an action and
+     * potentially a parameter with which to carry out the action.
+     *
+     * @param challenge the authentication challenge to handle
+     * @return the response that should be taken
+     */
+    @Override
+    public AuthenticationChallengeResponse handleChallenge(AuthenticationChallenge challenge) {
+
+      AuthenticationChallengeResponse ret = null;
+
+      if (challenge.getType() == AuthenticationChallenge.Type.OAUTH_CREDENTIAL_CHALLENGE) {
+        sOAuthChallengeCountDownLatch = new OAuthChallengeCountDownLatch(1);
+
+        final String portalUrl = challenge.getRemoteResource().getUri();
+
+        // get the OAuth config for the portalUrl associated with the challenge
+        try {
+          mOAuthConfiguration = AuthenticationManager.getOAuthConfiguration(portalUrl);
+        } catch (MalformedURLException e) {
+          return new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, challenge);
+        }
+        if (mOAuthConfiguration == null) {
+          return new AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, challenge);
```

Comment:
better to have just one return point at the end of the method

Reply:
updated